### PR TITLE
push tweaks 

### PR DESCRIPTION
**Pull request by zhili1208**
https://github.com/NuGet/Home/issues/4761

### CODE REVIEW

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 1]
```diff
@@ -24,6 +25,11 @@ namespace NuGet.Protocol.Core.Types
     public class PackageUpdateResource : INuGetResource
     {
         private const string ServiceEndpoint = "/api/v2/package";
+        private const string ApiKeyHeader = "X-NuGet-ApiKey";
+        private const string TempApiKeyServiceEndpoint = "create-verification-key/{0}/{1}";
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: nit: xmldoc of what the `{0}` and `{1}` parameters are

File: test/NuGet.Core.Tests/NuGet.Protocol.Core.v3.Tests/JsonData.cs [hunk 2]
```diff
@@ -4088,5 +4088,13 @@ public static class JsonData
           }";
         #endregion
 
+        #region TempApiKeyResponse
+        public const string tempApiKeyJsonData = @"{
+            ""Key"": ""tempkey"",
+            ""Expires"": ""/Date(1489106836905)/"",
```
Comment (joelverhagen) [inline test/NuGet.Core.Tests/NuGet.Protocol.Core.v3.Tests/JsonData.cs]: We should use ISO 6801 format instead of this.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 3]
```diff
@@ -526,5 +558,71 @@ private bool IsV2LocalRepository(string root)
...
+                        () =>
+                        {
+                            var request = HttpRequestMessageFactory.Create(
+                                HttpMethod.Put,
+                                serviceEndpointUrl,
+                                new HttpRequestMessageConfiguration(
+                                    logger: logger,
+                                    promptOn403: false));
+                            request.Headers.Add(ApiKeyHeader, apiKey);
+                            request.Headers.Add(NuGetClientVersionHeader, UserAgent.UserAgentVersion);
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: I think we should add the client version header to every HTTP call -- just like user agent.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 4]
```diff
@@ -526,5 +558,71 @@ private bool IsV2LocalRepository(string root)
...
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                return apiKey;
+            }
+            var serviceEndpointUrl = GetServiceEndpointUrl(NuGetConstants.DefaultGalleryServerUrl,
+                string.Format(TempApiKeyServiceEndpoint, packageIdentity.Id, packageIdentity.Version));
+
+            try
+            {
+                return await _httpSource.ProcessStreamAsync(
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: You can use `HttpSource.GetJObjectAsync` right? This will do the JSON parsing for you.

File: src/NuGet.Core/NuGet.Protocol.Core.Types/UserAgent.cs [hunk 5]
```diff
@@ -22,10 +22,13 @@ public static void SetUserAgentString(UserAgentStringBuilder builder)
             }
 
             UserAgentString = builder.Build();
+            UserAgentVersion = builder.NuGetClientVersion;
         }
 
         public static string UserAgentString { get; private set; }
 
+        public static string UserAgentVersion { get; private set; }
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/UserAgent.cs]: Use ``MinClientVersionUtility.GetNuGetClientVersion()`` instead.

This shouldn't be added here, it will just end up being a copy of the above property value which is cached already.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 6]
```diff
@@ -526,5 +560,53 @@ private bool IsV2LocalRepository(string root)
...
+                                    promptOn403: false));
+                            request.Headers.Add(ApiKeyHeader, apiKey);
+                            return request;
+                        }),
+                   logger,
+                   token);
+
+                return result.Value<string>("Key") ?? NullApiKey;
+            }
+            // If the package doesn't exist on nuget.org, nuget symbol server will ignore the api key.
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: Please put comments wither in the try or in the catch scope -- not in between

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 7]
```diff
@@ -57,18 +66,33 @@ public Uri SourceUri
...
                 var requestTimeout = TimeSpan.FromSeconds(timeoutInSecond);
                 tokenSource.CancelAfter(requestTimeout);
-
+                bool packageUseTempApiKey = false;
+                bool symbolUseTempApiKey = false;
                 var apiKey = getApiKey(_source);
 
-                await PushPackage(packagePath, _source, apiKey, requestTimeout, log, tokenSource.Token);
+                // If user push to https://nuget.smbsrc.net/ directly, use temp api key.
+                if (_source.Equals(NuGetConstants.DefaultSymbolServerUrl, StringComparison.OrdinalIgnoreCase))
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: Should this convert both to a Uri and try to normalize them, then compare the domain and path? I'm worried that this would miss variants such as ``https://nuget.smbsrc.net`` or ``http://nuget.smbsrc.net/`` or ``https://nuget.smbsrc.net/api/v2/packages`` or ``https://nuget.smbsrc.net//``

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 8]
```diff
@@ -57,18 +66,33 @@ public Uri SourceUri
...
                 // - The user specified it on the command line
                 // - The package source is nuget.org and the default symbol source is used
                 if (!string.IsNullOrEmpty(symbolSource))
                 {
                     string symbolApiKey = getSymbolApiKey(symbolSource);
-                    await PushSymbols(packagePath, symbolSource, symbolApiKey, requestTimeout, log, tokenSource.Token);
+
+                    // If symbol source is nuget.smbsr.net, always use temp key.
+                    // If package source is nuget.org and symbol api key is the same as package api key, use temp key
+                    if (symbolSource.Equals(NuGetConstants.DefaultSymbolServerUrl, StringComparison.OrdinalIgnoreCase)
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: Create a helper method for this check

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 9]
```diff
@@ -57,18 +66,33 @@ public Uri SourceUri
...
+                        || (IsSourceNuGetOrg() && apiKey == symbolApiKey))
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: Why does it matter if it is nuget.org? We discussed not special casing this I thought.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 10]
```diff
@@ -57,18 +66,33 @@ public Uri SourceUri
...
+                    {
+                        symbolUseTempApiKey = true;
+                    }
+                    await PushSymbols(packagePath, symbolSource, symbolApiKey, requestTimeout, log, symbolUseTempApiKey, tokenSource.Token);
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: I would create a helper method that does the check and push, that way this code will only exist in one place.

File: src/NuGet.Core/NuGet.Test.Utility/SimpleTestSetup/SimpleTestPackageUtility.cs [hunk 11]
```diff
@@ -38,14 +39,17 @@ public static class SimpleTestPackageUtility
            string repositoryDir,
            string id,
            string version,
-           IEnumerable<PackageDependency> dependencies)
+           IEnumerable<PackageDependency> dependencies,
+           bool isSymbolPackage = false)
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Test.Utility/SimpleTestSetup/SimpleTestPackageUtility.cs]: Add a new constructor instead of an optional parameter, as a generic type like this it gets painful

Comment (emgarten) [reply src/NuGet.Core/NuGet.Test.Utility/SimpleTestSetup/SimpleTestPackageUtility.cs]: You could also just make a CreateSymbolPackage method

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 12]
```diff
@@ -113,6 +137,7 @@ public Uri SourceUri
             string apiKey,
             TimeSpan requestTimeout,
             ILogger log,
+            bool useTempApiKey,
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: I would expect this to be decided at the end, passing it in provides more places for it to be wrong

**Review by emgarten — CHANGES_REQUESTED** [review]
Address feedback

Add tests for URL matching and negative scenarios to verify the key used

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 13]
```diff
@@ -526,5 +560,53 @@ private bool IsV2LocalRepository(string root)
...
+                            return request;
+                        }),
+                   logger,
+                   token);
+
+                return result.Value<string>("Key") ?? NullApiKey;
+            }
+            // If the package doesn't exist on nuget.org, nuget symbol server will ignore the api key.
+            // In this case, any api key should works, just set api key to some value here.
+            catch (HttpRequestException)
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: What sort of HTTP status codes will end up here? It seems like anything but 200 and 403 should be rejected.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs (hunk 7 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: Perhaps compare just on domain name? "nuget.smbsrc.net" part. It's possible the person types "https://nuget.smbsrc.net/api/v2/package" or "https://nuget.smbsrc.net/"

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs (hunk 13 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: 404 will not be rejected

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs (hunk 7 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: @zhili1208 mentioned there are some other scenarios/paths that this shouldn't be used for

**Review by emgarten — APPROVED** [review]
☘️

File: src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpRequestMessageFactory.cs [hunk 14]
```diff
@@ -99,6 +102,7 @@ public static HttpRequestMessage Create(HttpMethod method, Uri requestUri, ILogg
 
             var request = new HttpRequestMessage(method, requestUri);
             request.SetConfiguration(configuration);
+            request.Headers.Add(NuGetClientVersionHeader, MinClientVersionUtility.GetNuGetClientVersion().ToNormalizedString());
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpRequestMessageFactory.cs]: This should be set the same place as the user agent, right?

Comment (joelverhagen) [reply src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpRequestMessageFactory.cs]: https://github.com/NuGet/NuGet.Client/blob/af7f64c179055f741f6ed49b5d43ab6186ac6a1d/src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs#L316

File: test/NuGet.Core.Tests/NuGet.Protocol.Core.v3.Tests/JsonData.cs (hunk 2 above)
Comment (joelverhagen) [reply test/NuGet.Core.Tests/NuGet.Protocol.Core.v3.Tests/JsonData.cs]: I think Christy make the change for this -- can we modify the date to be:
`"2017-03-09T16:47:16"`

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 15]
```diff
@@ -24,6 +25,14 @@ namespace NuGet.Protocol.Core.Types
     public class PackageUpdateResource : INuGetResource
     {
         private const string ServiceEndpoint = "/api/v2/package";
+        private const string ApiKeyHeader = "X-NuGet-ApiKey";
+
+        /// <summary>
+        /// Create temporary verification api key endpoint: "create-verification-key/[package id]/[package version]"
+        /// </summary>
+        private const string TempApiKeyServiceEndpoint = "create-verification-key/{0}/{1}";
+        private const string NullApiKey = "ffffffff-0000-ffff-0000-ffffffffffff";
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: How did we pick this as the "null" API key? Why not all zeroes?

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 16]
```diff
@@ -526,5 +545,57 @@ private bool IsV2LocalRepository(string root)
...
+                return result.Value<string>("Key")?? InvalidApiKey;
+            }
+            catch(HttpRequestException ex)
+            {
+                if (ex.Message.Contains("Response status code does not indicate success: 403 (Forbidden)."))
+                {
+                    return InvalidApiKey;
+                }
+
+                throw ex;
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: just `throw;`

**Review by joelverhagen — APPROVED** [review]
Address comment then :shipit:

**Comment by emgarten** [comment]
:shipit:

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 17]
```diff
@@ -24,6 +25,13 @@ namespace NuGet.Protocol.Core.Types
     public class PackageUpdateResource : INuGetResource
     {
         private const string ServiceEndpoint = "/api/v2/package";
+        private const string ApiKeyHeader = "X-NuGet-ApiKey";
+        private const string InvalidApiKey = "invalidapikey";
+
+        /// <summary>
+        /// Create temporary verification api key endpoint: "create-verification-key/[package id]/[package version]"
+        /// </summary>
+        private const string TempApiKeyServiceEndpoint = "create-verification-key/{0}/{1}";
```
Comment (chenriksson) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs:34]: Per discussion today, need to remove version from create-verification-key endpoint

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 18]
```diff
@@ -526,5 +545,57 @@ private bool IsV2LocalRepository(string root)
...
+            string apiKey,
+            ILogger logger,
+            CancellationToken token)
+        {
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                return apiKey;
+            }
+            var serviceEndpointUrl = GetServiceEndpointUrl(NuGetConstants.DefaultGalleryServerUrl,
+                string.Format(TempApiKeyServiceEndpoint, packageIdentity.Id, packageIdentity.Version));
```
Comment (chenriksson) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs:614]: Reminder - remove version from create key endpoint

**Review by chenriksson — CHANGES_REQUESTED** [review]

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 19]
```diff
@@ -526,5 +545,57 @@ private bool IsV2LocalRepository(string root)
...
+
+            try
+            {
+                var result = await _httpSource.GetJObjectAsync(
+                    new HttpSourceRequest(
+                        () =>
+                        {
+                            var request = HttpRequestMessageFactory.Create(
+                                HttpMethod.Put,
```
Comment (chenriksson) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs]: Should change from PUT to POST, per https://github.com/NuGet/NuGetGallery/pull/3646#pullrequestreview-26707559

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs (hunk 18 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs:614]: per discussion today, client still send package version

**Review by joelverhagen — APPROVED** [review]

**Review by chenriksson — APPROVED** [review]

File: src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs [hunk 20]
```diff
@@ -526,5 +590,62 @@ private bool IsV2LocalRepository(string root)
...
+                        MaxTries = 1
+                    },
+                   logger,
+                   token);
+
+                return result.Value<string>("Key")?? InvalidApiKey;
+            }
+            catch(HttpRequestException ex)
+            {
+                if (ex.Message.Contains("Response status code does not indicate success: 403 (Forbidden)."))
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs:642]: Can the exception message be localized?  If I install a .NET Framework language pack, will this comparison fail?

Comment (zhili1208) [reply src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/PackageUpdateResource.cs:642]: Checked .net framework and .net core source code, this message is InvariantCulture
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L161
